packages/eslint-plugin-react-hooks/src/RulesOfHooks.js

```diff
@@ -571,59 +554,56 @@ export default {
-        const scope = context.getScope();
-        const variable = recursivelyFindVariableInScope(
-          scope,
-          node.callee.name,
-        );
-        if (variable) {
-          const isUseEvent = variable.defs.some(def => {
-            return isUseEventVariableDeclarator(def.node);
-          });
-          if (isUseEvent) {
-            let depth = 0;
-            for (const ancestor of context.getAncestors()) {
-              // For performance reasons we don't want to walk up
-              // every single ancestor node.
-              if (depth++ > MAX_USE_EVENT_ANCESTOR_DEPTH) {
-                break;
-              }
-              if (
-                ancestor.type === 'CallExpression' &&
-                ancestor.callee.name !== 'useEffect'
-              ) {
-                context.report({
-                  node: node.callee,
-                  message:
-                    'Functions created with React Hook "useEvent" must only be invoked in a ' +
-                    '"useEffect" or closure.',
-                });
-                break;
-              }
+        // useEvent: Resolve a function created with useEvent that is invoked locally at least once.
+        // OK - onClick();
+        resolveUseEventViolation(node.callee);
+
+        // useEvent: useEvent functions can be passed by reference within useEffect.
+        // OK - useEffect(() => { ...onClick... }, []);
+        if (node.callee.name === 'useEffect' && node.arguments.length > 0) {
+          // Subtraverse here so we don't need to visit every Identifier node.
+          traverse(context, node, path => {
+            if (path.node.type === 'Identifier') {
+              resolveUseEventViolation(path.node);
             }
-          }
+          });
         }
       },
 
-      JSXExpressionContainer(node) {
-        const scope = context.getScope();
-        // <Child onClick={onClick} />
-        if (node.expression.type === 'Identifier') {
-          reportOnUseEventDirectReferences(scope, node.expression);
+      MemberExpression(node) {
+        // useEvent: weird, but OK.
+        // OK - onClick.call(...);
+        // NOT OK - onClick.bind(...);
+        if (node.property.name === 'call' || node.property.name === 'apply') {
+          resolveUseEventViolation(node.object);
         }
+      },
 
-        // <Child onClick={onClick.bind(null)} />
-        if (
-          node.expression.type === 'CallExpression' &&
-          node.expression.callee.type === 'MemberExpression' &&
-          node.expression.callee.object.type === 'Identifier'
-        ) {
-          reportOnUseEventDirectReferences(
-            scope,
-            node.expression.callee.object,
-          );
+      Program(node) {
+        // useEvent: First pass to record all definitions of useEvent functions. This needs to run
+        // prior to the second pass so a subtraversal is necessary here.
+        traverse(context, node, path => {
```

Comment:
Can this traversal be done when we step inside a component rather than from the top level? And then when we’re inside a component, wouldn’t a shallow traversal be enough since all useEvent calls would be at the top level?

Reply:
Hmm, yeah we could, and I keep track of function scope as well. Great suggestion! I think it'd make the analysis more precise, since currently if you happen to have multiple components within a file, that happen to share the same variable name for the useEvent function, it won't catch that one might be uncalled.

Reply:
Interesting. I haven’t looked at the code closely but this can probably be solved by holding an identifier in a Map instead of a string? My concern was mostly from perf perspective — want to avoid adding any new deep traversals, and ideally fully ride on the existing traversal performed by the lint. 

Now that I think of it, I expect the “exhaustive deps” rule source to have all the logic already needed for this. Since we track things like “setters for state variables by identifier” there. Seems like “event functions by identifier” would fit well into that logic. Can we reuse it or follow the same approach in both for consistency?

Also now I’m thinking whether RulesOfHooks is the correct place for this rule. I’m kind of struggling to decide which one it belongs to. Almost feels like a new one.

Reply:
I'm inclined to ship this in RulesOfHooks now and separately follow up to break it apart into smaller rules.

Reply:
> the identifiers will have different SourceLocations so I'm not sure if they'll be equal

I think maybe in the other rule, we always find the original identifier and use that as a key? I’m not sure. But if there was an issue in that approach we should fix in both places regardless. I just want to avoid using different approaches for the same thing within one plugin.

Reply:
I updated it yesterday to drop the need for string keys now, it's just a plain Set of identifiers that we first collect then delete as the useEvent function is invoked. I added a test locally in ExhaustiveDeps, it doesn't have any issues with aliasing, and our approach now is consistent-ish with that rule. This one is just a lot simpler because we don't have to deal with figuring out dependencies and building dependency strings – we just check if the identifier the useEvent function was bound to was called at least once